torch/ao/quantization/fx/_decomposed.py

```diff
@@ -1203,3 +1203,45 @@ def convert_element_type(input: torch.Tensor, dtype: torch.dtype) -> torch.Tenso
 @impl(quantized_decomposed_lib, "convert_element_type.no_fuse", "Meta")
 def convert_element_type_meta(input: torch.Tensor, dtype: torch.dtype) -> torch.Tensor:
     return torch.empty_like(input, dtype=dtype)
+
+
+quantized_decomposed_lib.define(
+    "int4mm_packed_weight_cpu(Tensor input, Tensor weight, Tensor qGroupSize, Tensor scalesAndZeroPoints) -> Tensor"
+)
+
+
+@impl(quantized_decomposed_lib, "int4mm_packed_weight_cpu", "CPU")
+def int4mm_packed_cpu(
+    input: torch.Tensor,
+    weight: torch.Tensor,
+    qGroupSize: torch.Tensor,
+    scalesAndZeroPoints: torch.Tensor,
+) -> torch.Tensor:
+    """
+    A wrapper of `torch.ops.aten._weight_int4pack_mm_for_cpu` to support group size as a tensor.
+    Args:
+       input (torch.Tensor): bfloat16 activation Tensor
+       weight (torch.Tensor): int4 weights packed in int32 dtype Tensor
+       qGroupSize (torch.Tensor): group size for quantization
+       scaleAndZeroPoints (torch.tensor): scales & zero points in BF16.
+    Returns:
+       GEMM output
+    """
+    assert qGroupSize.numel() == 1, "qGroupSize must be a scalar"
+    assert qGroupSize.dtype == torch.int64, "qGroupSize must be int64"
```

Comment:
Hi, can you please also add an assert for `qgroupsize` being `(B.numel() * 2)/(scalesAndZeroPoints.numel() / 2)`?

We should probably also add such an assert in the existing C++ code for the int4 WoQ GEMM in https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/LinearAlgebra.cpp.

Thanks!

Reply:
Thanks for the comment. The check is the kernel's duty I think. Here we just need to check group size is a scalar tensor.

Reply:
> The check is the kernel's duty I think

Thanks, I agree. The ATen kernel currently isn't doing it, though. We should add a check there, perhaps in another PR.

Reply:
It's added.